### ISSUE TITLE
Additional Units: Conversion Rate based on Precision

### ISSUE THREAD

**Issue by frederic9583** [oroinc/orocommerce#101]
![image](https://user-images.githubusercontent.com/45210555/64233790-474c4180-cef5-11e9-90bd-e8e0613a53f3.png)

I'm wondering if this is right: If `item` has a `precision` of `0`, my additional unit `set` should not allow a `conversion_rate` of `10.33` but should get rounded to `10`, right? 

Do i miss something?

I also have a question regarding precision in combination with stock: If i'm adding a precision of `6`, how does orocommerce handle this within the inventory since we can only add two decimals in inventory (like minimum quantity etc)

Thanks!

**Comment by mbessolov** [comment]
@frederic9583 

> If item has a precision of 0, my additional unit set should not allow a conversion_rate of 10.33 but should get rounded to 10, right?

Not necessarily. For example, your primary unit may be **liter**, and an additional unit **gallon** would have a conversion rate of 3.785412 liters. Also, the primary unit is not necessarily the smallest unit, so a case with primary unit yard, and an additional unit of foot and conversion rate of 0.333333 should also be allowed.

> we can only add two decimals in inventory (like minimum quantity etc)

it's a bug and needs to be fixed

**Comment by frederic9583** [comment]
@mbessolov thanks for your feedback. I think there is a missunderstanding. I can't add add more than three decimals in the conversation rate in any case but also i can add some even if i don't want to.

## Example A:
- Kilogram with precision of `4`.
- Every additional unit should be allowed to have `4` decimals, right?
- Currently i'm only able to add three decimals:

![image](https://user-images.githubusercontent.com/45210555/64276093-61167480-cf47-11e9-8922-971754c1c701.png)

## Example B:
- Item with precision of `0`.
- Every additional unit should be restricted to have `0` decimals, right?
- Currently i'm able to add decimals to each additional unit:

![image](https://user-images.githubusercontent.com/45210555/64276188-9de26b80-cf47-11e9-80fd-e6da2ba1aeb2.png)

**Comment by mbessolov** [comment]
@frederic9583 

> I can't add add more than three decimals in the conversation rate in any case

OroCommerce by default (without customizing the installation process) can support up to 6 fractional digits, so any quanitty input in the backoffice or on the storefront should technically allow entry of up to 6 fractional digits as well. The actual number of fractional digits allowed for user entry should be limited by the "Precision" of that product unit.

The conversion rate is not really a quantity input, and it seems that at some point we limited the conversion rate value to 3 fractional digits. This limitation should be increased to at least the maximum default precision of 6 fractional digits, and possibly even more (e.g. 1 US gallon actually is 3.785411784 liters - 8 fractional digits necessary for more precise conversion). We'll see what can be done about it.

The reason not to tie the conversion rate precision to the quantity unit precision the conversion is not necessarily limited to "1 unit A to some whole number of unit B" even if you sell "unit B" as whole units. In my example with liters and gallons - you may want to sell only whole liters and only whole gallons, but the conversion rate from gallons to liters is 3.785411784 anyway.

**Comment by frederic9583** [comment]
Hey @mbessolov, thank you so much for your time!

> "you may want to sell only whole liters and only whole gallons, but the conversion rate from gallons to liters is 3.785411784 anyway."

PAM! Thanks for that - i just experienced a real enlightening! 😄 You may want to add that to your documentation, it's quite straight forward.

***

I also just checkout out the product prices and i think there is also a limitation of three fractional digits in the quantity. 

Here i defined an absurd precision of `9` for `set` and `0` for `item`.
![image](https://user-images.githubusercontent.com/45210555/64319859-9b225d80-cfbd-11e9-9542-1a7d5f700fac.png)

In product prices my `item` price gets trimmed to 0 fractional digits which is correct, but my `set` price should allow `9` digits, right?

![image](https://user-images.githubusercontent.com/45210555/64319841-93fb4f80-cfbd-11e9-8600-25e2a517fc34.png)

In the inventory, i'm able to add a stock for `set` with nine decimals.  
![image](https://user-images.githubusercontent.com/45210555/64319871-a2496b80-cfbd-11e9-8d0e-39d63242ac19.png)
